buildSrc/src/main/java/org/elasticsearch/gradle/testclusters/ElasticsearchCluster.java

```diff
@@ -0,0 +1,302 @@
+        }
+    }
+
+    @Override
+    public void setJavaHome(File javaHome) {
+        nodes.all(each -> each.setJavaHome(javaHome));
+    }
+
+    @Override
+    public void start() {
+        String nodeNames = nodes.stream().map(ElasticsearchNode::getName).collect(Collectors.joining(","));
+        for (ElasticsearchNode node : nodes) {
+            node.defaultConfig.put("cluster.name", safeName(clusterName));
+            if (Version.fromString(node.getVersion()).getMajor() >= 7) {
+                node.defaultConfig.put("cluster.initial_master_nodes", "[" + nodeNames + "]");
+                node.defaultConfig.put("discovery.seed_providers", "file");
+            }
+        }
+        nodes.forEach(ElasticsearchNode::start);
+    }
+
+    private void writeUnicastHostsFiles() {
+        String unicastUris = nodes.stream().flatMap(node -> node.getAllTransportPortURI().stream()).collect(Collectors.joining("\n"));
+        nodes.forEach(node -> {
+            try {
+                Files.write(node.getConfigDir().resolve("unicast_hosts.txt"), unicastUris.getBytes(StandardCharsets.UTF_8));
+            } catch (IOException e) {
+                throw new UncheckedIOException(e);
+            }
+        });
+    }
+
+    @Override
+    public String getHttpSocketURI() {
+        waitForAllConditions();
+        return getFirstNode().getHttpSocketURI();
+    }
+
+    @Override
+    public String getTransportPortURI() {
+        waitForAllConditions();
+        return getFirstNode().getTransportPortURI();
+    }
+
+    @Override
+    public List<String> getAllHttpSocketURI() {
+        waitForAllConditions();
+        return nodes.stream().flatMap(each -> each.getAllHttpSocketURI().stream()).collect(Collectors.toList());
+    }
+
+    @Override
+    public List<String> getAllTransportPortURI() {
+        waitForAllConditions();
+        return nodes.stream().flatMap(each -> each.getAllTransportPortURI().stream()).collect(Collectors.toList());
+    }
+
+    public void waitForAllConditions() {
+        long startedAt = System.currentTimeMillis();
+        logger.info("Waiting for nodes");
+        nodes.forEach(ElasticsearchNode::waitForAllConditions);
+
+        writeUnicastHostsFiles();
+
+        logger.info("Starting to wait for cluster to form");
+        addWaitForUri(
+            "cluster health yellow", "/_cluster/health?wait_for_nodes=>=" + nodes.size()  + "&wait_for_status=yellow"
+        );
+        waitForConditions(waitConditions, startedAt, CLUSTER_UP_TIMEOUT, CLUSTER_UP_TIMEOUT_UNIT, this);
+    }
+
+    @Override
+    public void stop(boolean tailLogs) {
+        nodes.forEach(each -> each.stop(tailLogs));
+    }
+
+    @Override
+    public boolean isProcessAlive() {
+        return ! nodes.stream().anyMatch(node -> node.isProcessAlive() == false);
+    }
+
+    public void eachVersionedDistribution(BiConsumer<String, Distribution> consumer) {
```

Comment:
This method seems pretty unnecessary. It simply loops over all nodes, and passes only the node version and distribution. I don't see why this is any better than just looping over the nodes themselves. This is tightly coupled to the corresponding use case in `TestClustersPlugin` and has no real reuse, and therfore should definitely not be `public`. I'd just replace the call in the plugin with a simple loop and remove this method.

Reply:
I made this package protected. 
The reason for the method is to be able to get this information about nodes without exposing everything about the nodes since we haven't decided on how to do that yet.

Reply:
I think it's the callback nature that seems unnecessary. We could just make this method return a set of unique version and distribution since that's _actually_ all we want.

Reply:
We can just leave this as-is for now.